packages/core/src/lib/stream_manager.ts

```diff
@@ -14,60 +18,108 @@ export class StreamManager {
     public addEventListener: Libp2p["addEventListener"]
   ) {
     this.log = new Logger(`stream-manager:${multicodec}`);
-    this.addEventListener(
-      "peer:update",
-      this.handlePeerUpdateStreamPool.bind(this)
-    );
-    this.getStream = this.getStream.bind(this);
     this.streamPool = new Map();
+
+    this.addEventListener("peer:update", this.handlePeerUpdateStreamPool);
+    this.getStream = this.getStream.bind(this);
   }
 
-  public async getStream(peer: Peer): Promise<Stream> {
+  public async getStream(peer: Peer): Promise<Stream | null> {
     const peerIdStr = peer.id.toString();
     const streamPromise = this.streamPool.get(peerIdStr);
 
     if (!streamPromise) {
-      return this.newStream(peer); // fallback by creating a new stream on the spot
+      return this.createStream(peer);
     }
 
-    // We have the stream, let's remove it from the map
     this.streamPool.delete(peerIdStr);
-
     this.prepareNewStream(peer);
 
-    const stream = await streamPromise;
-
-    if (!stream || stream.status === "closed") {
-      return this.newStream(peer); // fallback by creating a new stream on the spot
+    try {
+      const stream = await streamPromise;
+      if (stream && stream.status !== "closed") {
+        return stream;
+      }
+    } catch (error) {
+      this.log.error(`Failed to get stream for ${peerIdStr} -- `, error);
     }
 
-    return stream;
+    return this.createStream(peer);
+  }
+
+  private async createStream(peer: Peer): Promise<Stream | null> {
+    try {
+      return await this.newStream(peer);
+    } catch (error) {
+      this.log.error(
+        `Failed to create a new stream for ${peer.id.toString()} -- `,
+        error
+      );
+      return null;
+    }
   }
 
-  private async newStream(peer: Peer): Promise<Stream> {
+  private async newStream(peer: Peer, retries = 0): Promise<Stream> {
     const connections = this.getConnections(peer.id);
     const connection = selectConnection(connections);
+
     if (!connection) {
       throw new Error("Failed to get a connection to the peer");
     }
-    return connection.newStream(this.multicodec);
+
+    try {
+      return await connection.newStream(this.multicodec);
+    } catch (error) {
+      if (retries < MAX_RETRIES) {
+        const backoff = RETRY_BACKOFF_BASE * Math.pow(2, retries);
+        await new Promise((resolve) => setTimeout(resolve, backoff));
+        return this.newStream(peer, retries + 1);
+      }
+      throw error;
+    }
   }
 
   private prepareNewStream(peer: Peer): void {
-    const streamPromise = this.newStream(peer).catch(() => {
-      // No error thrown as this call is not triggered by the user
+    const timeoutPromise = new Promise<void>((resolve) =>
```

Comment:
let's make it a util or use some small npm module for it (I prefer to have some convenient module) - we can do a follow up